va/vampic.go

```diff
@@ -0,0 +1,427 @@
+	// metrics and log validation errors. Below here, do not use := to redeclare
+	// `prob`, or this will fail.
+	var prob *probs.ProblemDetails
+	var summary *mpicSummary
+	var localLatency time.Duration
+	start := va.clk.Now()
+	logEvent := validationLogEvent{
+		AuthzID:    req.Authz.Id,
+		Requester:  req.Authz.RegID,
+		Identifier: req.DnsName,
+		Challenge:  chall,
+	}
+	defer func() {
+		probType := ""
+		outcome := fail
+		if prob != nil {
+			probType = string(prob.Type)
+			logEvent.Error = prob.Error()
+			logEvent.Challenge.Error = prob
+			logEvent.Challenge.Status = core.StatusInvalid
+		} else {
+			logEvent.Challenge.Status = core.StatusValid
+			outcome = pass
+		}
+		// Observe local validation latency (primary|remote).
+		va.observeLatency(opChall, va.perspective, string(chall.Type), probType, outcome, localLatency)
+		if va.isPrimaryVA() {
+			// Observe total validation latency (primary+remote).
+			va.observeLatency(opChall, allPerspectives, string(chall.Type), probType, outcome, va.clk.Since(start))
+			logEvent.Summary = summary
+		}
+
+		// Log the total validation latency.
+		logEvent.Latency = va.clk.Since(start).Round(time.Millisecond).Seconds()
+		va.log.AuditObject("Validation result", logEvent)
+	}()
+
+	// Do local validation. Note that we process the result in a couple ways
+	// *before* checking whether it returned an error. These few checks are
+	// carefully written to ensure that they work whether the local validation
+	// was successful or not, and cannot themselves fail.
+	records, err := va.validateChallenge(
+		ctx,
+		identifier.NewDNS(req.DnsName),
+		chall.Type,
+		chall.Token,
+		req.ExpectedKeyAuthorization,
+	)
+
+	// Stop the clock for local validation latency.
+	localLatency = va.clk.Since(start)
+
+	// Check for malformed ValidationRecords
+	logEvent.Challenge.ValidationRecord = records
+	if err == nil && !logEvent.Challenge.RecordsSane() {
+		err = errors.New("records from local validation failed sanity check")
+	}
+
+	if err != nil {
+		logEvent.InternalError = err.Error()
+		prob = detailedError(err)
+		return bgrpc.ValidationResultToPB(records, filterProblemDetails(prob), va.perspective, va.rir)
+	}
+
+	if va.isPrimaryVA() {
+		// Do remote validation. We do this after local validation is complete to
+		// avoid wasting work when validation will fail anyway. This only returns a
+		// singular problem, because the remote VAs have already audit-logged their
+		// own validation records, and it's not helpful to present multiple large
+		// errors to the end user.
+		op := func(ctx context.Context, remoteva RemoteVA, req proto.Message) (remoteResult, error) {
+			validationRequest, ok := req.(*vapb.PerformValidationRequest)
+			if !ok {
+				return nil, fmt.Errorf("got type %T, want *vapb.PerformValidationRequest", req)
+			}
+			return remoteva.DoDCV(ctx, validationRequest)
+		}
+		summary, prob = va.doRemoteOperation(ctx, op, req)
```

Comment:
This is a bit of spooky action at a distance. Looking at this code, I thought "huh, that's weird; we drop `summary` on the ground after this. why doesn't a linter fire?"

And of course the answer is that on [line 267, in a `defer`](https://github.com/letsencrypt/boulder/pull/7870/files#diff-5d6a394d08b52f99a8d1f8c60f1c89495ddb3f54c6e1cf1322d43d819f7130adR267), we do:

```
logEvent.Summary = summary
```

Since this is the only place that `summary` can possibly be set, we could reduce a level of indirection, and set `logEvent.Summary` right here.

On the other hand, I see that we already follow a similar pattern for `prob`, so there's an argument to be made for treating this as local style, and doing a cleanup pass later (I think in general we are doing too much in the `defer` and could pull a lot out into more straight-line code). What do you think?